### ISSUE TITLE
Bug generating Bluesky post facets when text includes `[` `]` brackets

### ISSUE THREAD

**Issue by snarfed** [snarfed/bridgy-fed#1605]
Original:
* https://flipboard.social/@TechDesk/113595595024808829
* https://browser.pub/https://flipboard.social/@TechDesk/113595595024808829

Bridged:
* https://bsky.app/profile/TechDesk.flipboard.social.ap.brid.gy/post/3lcimk7zy2gy2
* https://atproto-browser.vercel.app/at/TechDesk.flipboard.social.ap.brid.gy/app.bsky.feed.post/3lcimk7zy2gy2

<img width="400" alt="image" src="https://github.com/user-attachments/assets/cc77be6f-3483-43f7-b635-64650ebad83a">
<img width="400" alt="image" src="https://github.com/user-attachments/assets/a890b1e0-2427-41fa-b253-fc9141b70793">

**Comment by Tamschi** [comment]
That's two bugs in one. Any Markdown control characters need to be escaped (because otherwise it'll break for posts that contain e.g. `[a](https://example.com/)` as _plain_ text), and it seems the Markdown-to-Bluesky conversion doesn't behave like most Markdown processors would.

**Comment by snarfed** [comment]
> it seems the Markdown-to-Bluesky conversion doesn't behave like most Markdown processors would

Hmm! We don't currently expect or parse any external data as Markdown afaik, our use of Markdown is entirely a (small) internal implementation detail that shouldn't cause externally-visible bugs like this.

**Comment by Tamschi** [comment]
> > it seems the Markdown-to-Bluesky conversion doesn't behave like most Markdown processors would
> 
> Hmm! We don't currently expect or parse any external data as Markdown afaik, our use of Markdown is entirely a (small) internal implementation detail that shouldn't cause externally-visible bugs like this.

No, Bridgy Fed does in fact parse _all_ external content as Markdown right now since it doesn't escape Markdown control characters.
As I wrote, doing so will likely hide irregularities with the implementation though, as long as it handles those escape sequences well.

This is about the same issue as #1380 where no HTML escape is performed when converting to HTML for ActivityPub.

**Comment by snarfed** [comment]
Oh, I thought you meant u_deliberately_ parse Markdown, as in expect it anywhere inbound. Unintentionally, on the other hand, yes. 😁

**Comment by snarfed** [comment]
Related: #1144

**Comment by snarfed** [comment]
I hoped I could fix this by extracting links and converting them to facets _before_ [calling `html_to_markdown` here](https://github.com/snarfed/granary/blob/1a15c241b8ccf158cff55194c11d87cf1644378d/granary/bluesky.py#L723-L732), but `html_to_markdown` returns entirely new text, so the indices would be off.

I tried changing [`MARKDOWN_LINK_RE`](https://github.com/snarfed/granary/blob/1a15c241b8ccf158cff55194c11d87cf1644378d/granary/bluesky.py#L164) to start with `r'\[(?P<text>[^]]*?)\]`, which is less greedy, and works, but regresses granary's `test_bluesky.BlueskyTest.test_from_as1_html_link`.

**Comment by snarfed** [comment]
...and `html2text` doesn't escape `]` characters in its output Markdown link text, which is what `test_bluesky.BlueskyTest.test_from_as1_html_link` tests.

Maybe I can reuse these, from `html2text.config`:

```py
RE_SLASH_CHARS = r"\`*_{}[]()#+-.!"
RE_MD_BACKSLASH_MATCHER = re.compile(
    r"""
    (\\)          # match one slash
    (?=[%s])      # followed by a char that requires escaping
    """
    % re.escape(RE_SLASH_CHARS),
    flags=re.VERBOSE,
)
```

**Comment by snarfed** [comment]
Fixed!
